src/jmh/java/software/amazon/cloudwatchlogs/emf/MetricsLoggerBenchmark.java

```diff
@@ -0,0 +1,349 @@
+     * 32.159; S lock: 34.226
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    public void measureAllMethodsWith10Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(10);
+    }
+
+    /**
+     * Each thread executes all methods 100 times, 20 threads in total. no lock (need to sync
+     * getAllDimensions() & getAllDimensionKeys() in MetricsDirective): 11.833 ms/op; RW lock:
+     * 60.510 ms/op; S lock: 75.125 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    public void measureAllMethodsWith20Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(20);
+    }
+
+    /**
+     * Each thread executes all methods 100 times, 50 threads in total. no lock (need to sync
+     * getAllDimensions() & getAllDimensionKeys() in MetricsDirective): 36.051 ms/op; RW lock:
+     * 150.022 ms/op; S lock: 244.934 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    public void measureAllMethodsWith50Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(50);
+    }
+
+    /**
+     * Each thread executes all methods 100 times, 200 threads in total. no lock (need to sync
+     * getAllDimensions() & getAllDimensionKeys() in MetricsDirective): 108.775 ms/op; RW lock:
+     * 629.826 ms/op; S lock: 1220.959 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    @Warmup(time = 10)
+    @Measurement(time = 10)
+    public void measureAllMethodsWith200Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(200);
+    }
+
+    /**
+     * Each thread executes all methods 100 times, 500 threads in total. no lock (need to sync
+     * getAllDimensions() & getAllDimensionKeys() in MetricsDirective): 335.183 ms/op; RW lock:
+     * 1741.003 ms/op; S lock: 4192.327 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    @Warmup(time = 10)
+    @Measurement(time = 10)
+    public void measureAllMethodsWith500Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(500);
+    }
+
+    /**
+     * Each thread executes all methods 100 times, 1000 threads in total. no lock (need to sync
+     * getAllDimensions() & getAllDimensionKeys() in MetricsDirective): 575.339 ms/op; RW lock:
+     * 3230.403 ms/op; S lock: 13519.459 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    @Warmup(time = 20)
+    @Measurement(time = 20)
+    public void measureAllMethodsWith1000Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(1000);
+    }
+
+    private void measurePutMetricWithNThreads(int n) throws InterruptedException {
+        logger = new MetricsLogger(envProvider);
+        Thread[] threads = new Thread[n];
+        int batchSize = 1000;
+
+        for (int i = 0; i < n; i++) {
+            final int id = i;
```

Comment:
I don't think we need this separate variable. How about just renaming `i` to `id` or `threadId`?